### PR TITLE
mgr/volumes: subvoulme snapshot mirroring interface

### PR DESCRIPTION
**Pull request by vshankar**
Keeping this draft atm.

The idea is to introduce mirroring interfaces in mgr/volumes to synchronize subvolume snapshots. The approach is pretty straightforward, however, we probably need to discuss more about accessibility of the synchronized snapshots via `ceph fs subvolume ...` based mgr interfaces.

Since snapshots are synchronized by copying snap data to the remote file system followed by taking a snapshot of the remote directory, there is a possibility of an interrupted sync (mirror daemon terminating, etc..) which could lead to the remote directory content half-baked. This can cause subvolume based commands on the remote file system not function as expected (think if the .meta is left in a half-baked state when the mirror daemon is interrupted). I.e., the "remote" subvolume is in an inconsistent state till the mirror daemon snaps the remote directory.

Signed-off-by: Venky Shankar <vshankar@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/volumes/fs/volume.py [hunk 1]
```diff
@@ -379,6 +380,69 @@ def list_subvolumes(self, **kwargs):
...
+        volname = kwargs["vol_name"]
+        subvolname = kwargs["sub_name"]
+        groupname = kwargs["group_name"]
+
+        try:
+            with open_volume(self, volname) as fs_handle:
+                with open_group(fs_handle, self.volspec, groupname) as group:
+                    with open_subvol(self.mgr, fs_handle, self.volspec, group, subvolname, SubvolumeOpType.GETPATH) as subvolume:
+                        snap_path = subvolume.snapshot_base_path().decode("utf-8")
+                        subvolume.set_mirrored()
```
Comment (vshankar) [inline src/pybind/mgr/volumes/fs/volume.py:394]: FYI, this will be used to guard subvolume removal (and can be overridden by `--force`)

**Comment by batrick** [comment]
Please create a tracker ticket @vshankar

**Comment by vshankar** [comment]
> Please create a tracker ticket @vshankar

There's already one: https://tracker.ceph.com/issues/48944 (didn't link it since this was a draft pr)

File: src/pybind/mgr/volumes/module.py [hunk 2]
```diff
@@ -320,6 +320,24 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
             'desc': "Cancel an pending or ongoing clone operation.",
             'perm': 'r'
         },
+        {
+            'cmd': 'fs subvolume snapshot mirror enable '
```
Comment (ajarr) [inline src/pybind/mgr/volumes/module.py:324]: You expect these 2 commands to be used instead of `fs snapshot mirror add/remove` going forward?

**Review by ajarr — COMMENTED** [review]
> Since snapshots are synchronized by copying snap data to the remote file system followed by taking a snapshot of the remote directory, there is a possibility of an interrupted sync (mirror daemon terminating, etc..) which could lead to the remote directory content half-baked. This can cause subvolume based commands on the remote file system not function as expected (think if the .meta is left in a half-baked state when the mirror daemon is interrupted). I.e., the "remote" subvolume is in an inconsistent state till the mirror daemon snaps the remote directory.

Are the 2 interfaces you added in this draft affected by the issue discussed above? It doesn't seem like it.  What  volumes module's interfaces are going to be affected by the above issue? I don't see how the issue discussed above is related to the tracker ticket https://tracker.ceph.com/issues/48944 , which is to find an admin friendly interface replacement/enhancement for  "fs snapshot mirror add/remove".

**Comment by vshankar** [comment]
> Are the 2 interfaces you added in this draft affected by the issue discussed above? It doesn't seem like it. What volumes module's interfaces are going to be affected by the above issue? I don't see how the issue discussed above is related to the tracker ticket https://tracker.ceph.com/issues/48944 , which is to find an admin friendly interface replacement/enhancement for "fs snapshot mirror add/remove".

So, the idea used in the PR is to approach the issue from the "other side". I.e., rather than introducing interfaces in mgr/mirroring to accept subvolume/group, mgr/volumes provides the "mirroring" interface (since it already accepts subvolume/group) and uses mgr/mirroring to schedule snap synchronization for subvolume snapshots.

The issue mentioned here: https://github.com/ceph/ceph/pull/42091#issue-679769117, is the general problem of using `fs subvolume ...` commands on the remote file system when snap sync is in-progress. It has less to do with whichever way the interfaces are introduced.

**Comment by ajarr** [comment]
> > Are the 2 interfaces you added in this draft affected by the issue discussed above? It doesn't seem like it. What volumes module's interfaces are going to be affected by the above issue? I don't see how the issue discussed above is related to the tracker ticket https://tracker.ceph.com/issues/48944 , which is to find an admin friendly interface replacement/enhancement for "fs snapshot mirror add/remove".
> 
> So, the idea used in the PR is to approach the issue from the "other side". I.e., rather than introducing interfaces in mgr/mirroring to accept subvolume/group, mgr/volumes provides the "mirroring" interface (since it already accepts subvolume/group) and uses mgr/mirroring to schedule snap synchronization for subvolume snapshots.
> 

This I was able to infer from the code.

> The issue mentioned here: [#42091 (comment)](https://github.com/ceph/ceph/pull/42091#issue-679769117), is the general problem of using `fs subvolume ...` commands on the remote file system when snap sync is in-progress. It has less to do with whichever way the interfaces are introduced.

OK. I want to understand the scope of this PR.  There is no commit message. Is it to solve the tracker ticket https://tracker.ceph.com/issues/48944 or  "The idea is to introduce mirroring interfaces in mgr/volumes to synchronize subvolume snapshots. The approach is pretty straightforward, however, we probably need to discuss more about accessibility of the synchronized snapshots via ceph fs subvolume ... based mgr interfaces." ?

**Comment by vshankar** [comment]
> OK. I want to understand the scope of this PR. There is no commit message. Is it to implement a replacement API for `fs snapshot mirror add/remove` , or solve a bigger underlying issue?

I kind of explained what this PR is for in the initial comment. It is (obviously) to fix the tracker in question.

So, there are a couple of things broadly:

1. RIght now, the way subvolume snapshots can be mirrored is through adding the subvolume snapshot path via `fs  snapshot mirror add ...` interface. Since this error prone as you know since the snapshot path depends on the subvolume version. So, we need a mgr cmd interface that accepts subvolume/group and "internally" schedule mirroring for its snapshots.
2. The way mirror daemon synchronizes snapshots results in subvolume family of commands (`fs subvolume ...`) to be inoperable on the remote file system when a snap sync is in-progress. Is this acceptable to the users? Maybe or maybe not -- I don't know. It is not a good user experience imo. And this is what _requires_ discussion (w/ CSI folks, etc..).

As far as the scope of this PR -- Users should be able to schedule snapshot mirroring for subvolumes and not go through the tedious process of fetching the subvolume path, figuring the snapshot path, and finally adding that path for mirroring. Note that the issue mentioned in point#2 above applies here too.

Depending on what's acceptable to users, we may need to change the way we synchronize snapshots or do some gymnastics in the MDS to workaround point#2.

Does this make things clear?

**Comment by ShyamsundarR** [comment]
@madhu-1 take a look at the interface provided. Thanks!

**Comment by ajarr** [comment]
> > OK. I want to understand the scope of this PR. There is no commit message. Is it to implement a replacement API for `fs snapshot mirror add/remove` , or solve a bigger underlying issue?
> 
> I kind of explained what this PR is for in the initial comment. It is (obviously) to fix the tracker in question.
> 
> So, there are a couple of things broadly:
> 
>     1. RIght now, the way subvolume snapshots can be mirrored is through adding the subvolume snapshot path via `fs  snapshot mirror add ...` interface. Since this error prone as you know since the snapshot path depends on the subvolume version. So, we need a mgr cmd interface that accepts subvolume/group and "internally" schedule mirroring for its snapshots.
> 
>     2. The way mirror daemon synchronizes snapshots results in subvolume family of commands (`fs subvolume ...`) to be inoperable on the remote file system when a snap sync is in-progress. Is this acceptable to the users? Maybe or maybe not -- I don't know. It is not a good user experience imo. And this is what _requires_ discussion (w/ CSI folks, etc..).
> 
> 
> As far as the scope of this PR -- Users should be able to schedule snapshot mirroring for subvolumes and not go through the tedious process of fetching the subvolume path, figuring the snapshot path, and finally adding that path for mirroring. 

This makes sense to me.

> Note that the issue mentioned in point#2 above applies here too.
> 

This is one isn't clear to me. I understand that there is an issue of mgr volume commands not working on remote/secondary file system during in-progress snap sync . But the command you're implementing here is directed to the primary file system and should not be concerned with the remote/secondary file system, right?

> Depending on what's acceptable to users, we may need to change the way we synchronize snapshots or do some gymnastics in the MDS to workaround point#2.
> 
> Does this make things clear?

**Comment by vshankar** [comment]
> This is one isn't clear to me. I understand that there is an issue of mgr volume commands not working on remote/secondary file system during in-progress snap sync . But the command you're implementing here is directed to the primary file system and should not be concerned with the remote/secondary file system, right?

Right. The command itself isn't concerned about the remote file system. However, how does the user access snapshots on the remote file system via subvolume based commands? Imagine the user wanting to switch over to the remote cluster (hence remote file system) since the primary cluster has borked for some reason. Now, the user tries to restore a snap (or clone a snapshot). How will these operations work on the remote file system iff the metadata is not fully up-to-date? (maybe the mirror daemon was stopped when it was in-midst of syncing some snapshot).

**Comment by ajarr** [comment]
> 
> Right. The command itself isn't concerned about the remote file system. However, how does the user access snapshots on the remote file system via subvolume based commands? Imagine the user wanting to switch over to the remote cluster (hence remote file system) since the primary cluster has borked for some reason. Now, the user tries to restore a snap (or clone a snapshot). How will these operations work on the remote file system iff the metadata is not fully up-to-date? (maybe the mirror daemon was stopped when it was in-midst of syncing some snapshot).

Agree. This is a harder problem to solve and we need to discuss this. Thanks!

**Comment by ShyamsundarR** [comment]
> > This is one isn't clear to me. I understand that there is an issue of mgr volume commands not working on remote/secondary file system during in-progress snap sync . But the command you're implementing here is directed to the primary file system and should not be concerned with the remote/secondary file system, right?
> 
> Right. The command itself isn't concerned about the remote file system. However, how does the user access snapshots on the remote file system via subvolume based commands? Imagine the user wanting to switch over to the remote cluster (hence remote file system) since the primary cluster has borked for some reason. Now, the user tries to restore a snap (or clone a snapshot). How will these operations work on the remote file system iff the metadata is not fully up-to-date? (maybe the mirror daemon was stopped when it was in-midst of syncing some snapshot).

@vshankar Any partly synced snapshot needs to be rolled back, for any sub-directory that is mirrored. This includes the subvolume as well, so not necessarily a special case that needs handling as I see it. I think we discussed this earlier, and the answer was to roll back to an earlier fully synced snapshot if the mirror is promoted as primary, is this the prescribed method?

Also, the mirror being unusable, in terms of subvolume operations, would not be acceptable. The way it should work is when promoted, internally Ceph should decide if the mirror is fully synced copy or not, and take actions to either prevent the promotion as primary or rectify it internally and then promote.

**Comment by vshankar** [comment]
> @vshankar Any partly synced snapshot needs to be rolled back, for any sub-directory that is mirrored. This includes the subvolume as well, so not necessarily a special case that needs handling as I see it. I think we discussed this earlier, and the answer was to roll back to an earlier fully synced snapshot if the mirror is promoted as primary, is this the prescribed method?
> 
> Also, the mirror being unusable, in terms of subvolume operations, would not be acceptable. The way it should work is when promoted, internally Ceph should decide if the mirror is fully synced copy or not, and take actions to either prevent the promotion as primary or rectify it internally and then promote.

@ShyamsundarR I'll get to this once I'm done with other (priority) work.

I understand your point but I do not have anything right now to take the discussion forward.

**Pull request closed without merging**